Place HKDF info in to a temporary buffer if required.

vcsjones opened the pull request:
If the info input overlaps with the output destination, copy
the info to a temporary buffer first. As the HKDF expansion
is performed, the output buffer is filled by hash size. However
this will overwrite the info, should the buffers overlap.

The other parameters, like prk, salt, and ikm do not need any addional
handling for overlaps.

1. `Extract` is basically an HMAC. The `salt` is the key (and the underlying `HMAC` implementations have their own "owned" key material handled by the underlying crypto provider). The input keying material is appended to the digest. Even if the buffers for these overlap with the pseudo random key, it works. Tests were added nevertheless.
2. `Expand`'s PRK is the HMAC key, so it behaves as described above. No copying required.
3. `Info` is the only parameter where we are writing and reading to it at the same time that will affect the operation.

Easier to review with "ignore white space".

Closes #47009

vcsjones committed:
Place HKDF info in to a temporary buffer if required.

If the info input overlaps with the output destination, copy
the info to a temporary buffer first. As the HKDF expansion
is performed, the output buffer is filled by hash size. However
this will overwrite the info, should the buffers overlap.

The other parameters, like prk, salt, and ikm do not need any addional
handling for overlaps.

krwq commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/HKDF.cs:
```diff
@@ -134,35 +134,66 @@ private static void Expand(HashAlgorithmName hashAlgorithmName, int hashLength,
-                    hmac.AppendData(counterSpan);
+                tempInfoBuffer = info.Length > MaxStackInfoBuffer ?
+                    (rentedTempInfoBuffer = CryptoPool.Rent(info.Length)).AsSpan(0, info.Length) :
+                    (stackalloc byte[MaxStackInfoBuffer]).Slice(0, info.Length);
```
I'm a bit confused about this stackalloc - usually you get stack allocated buffer back when you get out of the current scope which in this case is `}` from the `if` statement which would mean when buffer overlaps you always end up writing where you shouldn't be.

At the same time the variable scope is different so I'm not certain this will work as expected.

cc: @GrabYourPitchforks

bartonjs replied:
> usually you get stack allocated buffer back (esp) when you get out of the current scope

Nope, end of the method.  It's why stackalloc in a loop is bad.

```C#
for (int i = 0; i < 1024; i++)
{
    Span<byte> tmp = stackalloc byte[256];
}
```

That's a 256kb stackalloc, split into tiny pieces.

vcsjones replied:
Oh, yikes. 🤦 thanks for catching that.

vcsjones replied:
Oh, wait - I thought you were saying I was `stackalloc`-ing in a loop, but that does not seem to be the case?

bartonjs replied:
I was just giving a counter-example to @krwq 's assertion that stackallocs are popped on block boundaries.  (I guess the easier one would have been that all code like

```C#
Span<byte> something = ...;

if (somethingBoolean)
{
    something = stackalloc byte[20];
}

// use something
```

would be broken)

krwq commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/HKDF.cs:
```diff
@@ -134,35 +134,66 @@ private static void Expand(HashAlgorithmName hashAlgorithmName, int hashLength,
             }
+            finally
+            {
+                CryptographicOperations.ZeroMemory(tempInfoBuffer);
```
isn't cryptopool already zeroing bytes?

vcsjones replied:
The `CryptoPool` `Return` below does not do the zero because `tempInfoBuffer` might be stack data, or it might be pool data. I erred on the size of zeroing the data regardless of which since we don't know what developers might be feeding in to the `info` parameter. Since we have the `ZeroMemory`, we don't need to clear it again when returning to the pool.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/HKDF.cs:
```diff
@@ -134,35 +134,66 @@ private static void Expand(HashAlgorithmName hashAlgorithmName, int hashLength,
             Span<byte> remainingOutput = output;
 
-            using (IncrementalHash hmac = IncrementalHash.CreateHMAC(hashAlgorithmName, prk))
+            ReadOnlySpan<byte> infoBuffer = stackalloc byte[0];
```
While this looks like .NET Core 2.0 Jeremy code, I feel like .NET 6 Jeremy code would just always make the stackalloc.

```C#
Span<byte> tempInfoBuffer = stackalloc byte[64];
ReadOnlySpan<byte> infoBuffer = tempInfoBuffer;
byte[]? rentedTempInfoBuffer = null;

if (overlaps)
{
    if (info.Length > tempInfoBuffer.Length)
    {
        rented = Rent();
        tempInfoBuffer = rented;
    }

    tempInfoBuffer = tempInfoBuffer.Slice(0, info.Length);
    info.CopyTo(tempInfoBuffer);
}
else
{
    infoBuffer = info;
}
```

vcsjones replied:
Hm. That's fair. I like the readability of this more.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/HKDF.cs:
```diff
@@ -134,35 +134,66 @@ private static void Expand(HashAlgorithmName hashAlgorithmName, int hashLength,
             }
+            finally
+            {
+                CryptographicOperations.ZeroMemory(tempInfoBuffer);
```
I'm having trouble convincing myself that `info` should be given an aggressive erasure (e.g. even a stack erasure).

Once it's not worthy of stack erasure, we can get rid of the try/finally and then just only return on success.  (Clearing on rent-return is still good)

vcsjones replied:
Fair enough. Do you have thoughts on what would warrant clearing stack data? Symmetric / private keys only?

bartonjs replied:
I think my neural net maps as: temporary copies of symmetric keys, private keys, data that is going into encrypt and data that is coming out of decrypt.

There might be other cases, or more nuance.

vcsjones committed:
Code review feedback.

* Make stackalloc code more readable by always allocating on the stack
up-front.
* Less-aggressively clear stack data, only clear when returning to pool.
    don't return to pool in a finally - let it get GCed if there is an
    exception.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/tests/HKDFTests.cs:
```diff
@@ -549,6 +549,78 @@ public void Rfc5869DeriveKeySpanOkmPotentiallyOverflowingValue()
+                info.CopyTo(overlap);
+                info = overlap;
+
+                HKDF.Expand(HashAlgorithmName.SHA256, prk, output: overlap, info: info);
```
I don't see a test that also verifies that prk/output can overlap. (the only pair missing)

All of these are overlapping perfectly.  Should we also test overlap forward and overlap backwards?

vcsjones replied:
>  don't see a test that also verifies that prk/output can overlap.

:thinking: Hm. Fixed.

> Should we also test overlap forward and overlap backwards?

I kind of waffled on this. I initially went with "no" since I could prove it to myself with assumptions about other tests and looking through the code, but on the otherhand, "why not" so that it is clear to folks that it is expected to work without thinking about it too much. So I added tests for that.

vcsjones committed:
Add test for overlapping PRK and output for Expand

vcsjones committed:
Add tests for overlapping buffers with various offsets.
